config/initializers/sidekiq.rb

```diff
@@ -6,8 +6,11 @@
   # Set the network timeout to be 5 seconds instead of the default 1 to help with
   # cloud network latency and longer batch commands
   config.redis = { url: ENV['REDIS_URL'], network_timeout: 5 }
+  Sidekiq::Status.configure_server_middleware config, expiration: 30.minutes
+  Sidekiq::Status.configure_client_middleware config, expiration: 30.minutes
```

Comment:
Can you explain exactly what these are doing?

Reply:
The `configure_server_middleware` method registers the status middleware which handles storing the actual status of the job as it moves from `:queued` to `:working` to `:complete`. The `expiration` sets how long this data is stored in redis, so after 30 mins any stuff that `sidekiq-status` stores will be wiped away. The `configure_client_middleware` stores the initial status of the job as `:queued` and associates that with the `jid` of the job. In both cases, this middleware processing is bypassed unless `Sidekiq::Status::Worker` is included in the job class. So none of this should happen for anything outside of the `ExportFhirJob`. The client middleware is configured for the server and client because in theory if a job spawned another job, that first job would be acting as a client.

That's the summary of my understanding of what this is up to, but if you want to look through anything yourself, here are the links I was using to figure this out:
* Server middleware code: https://github.com/kenaniah/sidekiq-status/blob/main/lib/sidekiq-status/server_middleware.rb
* Client middleware code: https://github.com/kenaniah/sidekiq-status/blob/main/lib/sidekiq-status/client_middleware.rb
* Sidekiq status config docs: https://github.com/kenaniah/sidekiq-status#configuration
* Sidekiq middleware docs: https://github.com/mperham/sidekiq/wiki/Middleware

Reply:
I attempted to stress test exporting on a 500K patient export (the performance database with root jurisdiction). At some point, something must have gone wrong, but it's not obvious what. After showing me that the job was in progress for a while, I now get a 404 when I hit the `ExportStatus` endpoint, even though the sidekiq job is still busy. My best guess is that this happened at the 30 minute mark due to the limit imposed here.

Was 30 minutes mostly arbitrary? I'm wondering if we should increase the expiration. Do you have a sense of how big of an export would take us to that limit?

Based on what I observed, I'm also thinking hitting that limit should result in the job being terminated (I haven't looked into this though). And also ideally at that point, the `ExportStatus` endpoint would return a 500 with Response Type of [Error](https://hl7.org/fhir/uv/bulkdata/export/index.html#endpoint-1) rather than a 404.

Reply:
The 30 minutes was relatively arbitrary, so that number is flexible. Based on my testing, an export of about 40,000 is going to bump up against that limit. Something like 30,000 would be safe, and something like 50,000 likely a bit over the limit. Caveat that this is just me testing locally, but I think these rough estimates will work.

I've also confirmed that your guess is right, when the expiration is surpassed, you get that 404. It is just an issue with the status, and if you wait for the job to complete, you can actually still access the created files.

So the question then is how long do we want to make this expiration time. Again in my testing (and just from what we can expect from how the code works), the export time should scale approx linearly with number of patients. So for 100K patients, we'd be looking at ~75 minutes. For 500K, it would be ~6.5 hrs. Any thoughts on what a good expiration time might be? We can just be conservative and make it something like 12 hours, but that does allow for quite long running `ExportFhirJob` runs. I don't imagine that leaving the minor amount of metadata stored for that long would cause an issue (unless the number of exports different jurisdictions are doing per 12 hours got very large, but that seems less likely to me)

Reply:
I wouldn't be concerned about the metadata, but if we allow for jobs to run too long, we theoretically open up the possibility of consuming all the export worker threads with `ExportFhirJob`s, e.g. if a client were to kick off 200K patient exports every 15 mins (the retry interval), which would affect regular exports as well.

The reality though is that we don't have any API clients for jurisdictions with enough monitorees for this to be an issue, so I don't think we should overthink it right now. Maybe let's just use 60 mins, which would mean we consume an absolute maximum of 4 worker threads at a time.

Although also this still leaves the open question of can we also abort the job when we hit the expiration? Do you know if that's possible?

Reply:
Alright I will bump to 60. I haven't done the implementing yet, but I think I should be able to cancel the job with something like this: https://github.com/mperham/sidekiq/wiki/FAQ#how-do-i-cancel-a-sidekiq-job.

Reply:
Ok so https://github.com/SaraAlert/SaraAlert/pull/987/commits/fef5e9e9487bf3f37b213141467c2b2ae42fff10 handles this, but it didn't quite work as I had thought it could. That link I sent in my above message really only lets you unschedule a job, not cancel a job that is already running. So I've had to leave it up to the job itself to check if it has been running too long and quit if so in https://github.com/SaraAlert/SaraAlert/pull/987/commits/fef5e9e9487bf3f37b213141467c2b2ae42fff10. One thing that is slightly annoying but I couldn't figure out a good way around is that it is impossible on the `APIController` side when checking status to know if a job just doesn't exist, or did exist but ran for too long and was aborted.